xds/src/test/java/io/grpc/xds/FakeControlPlaneXdsIntegrationTest.java

```diff
@@ -0,0 +1,333 @@
+    HttpFilter routerFilter = HttpFilter.newBuilder()
+        .setName("terminal-filter")
+        .setTypedConfig(
+            Any.pack(Router.newBuilder().build()))
+        .setIsOptional(true)
+        .build();
+    VirtualHost virtualHost = io.envoyproxy.envoy.config.route.v3.VirtualHost.newBuilder()
+        .setName("virtual-host-0")
+        .addDomains(authority)
+        .addRoutes(
+            Route.newBuilder()
+                .setMatch(
+                    RouteMatch.newBuilder().setPrefix("/").build())
+                .setNonForwardingAction(NonForwardingAction.newBuilder().build())
+                .build()).build();
+    RouteConfiguration routeConfig = RouteConfiguration.newBuilder()
+        .addVirtualHosts(virtualHost)
+        .build();
+    Filter filter = Filter.newBuilder()
+        .setName("network-filter-0")
+        .setTypedConfig(
+            Any.pack(
+                HttpConnectionManager.newBuilder()
+                    .setRouteConfig(routeConfig)
+                    .addAllHttpFilters(Collections.singletonList(routerFilter))
+                    .build())).build();
+    FilterChainMatch filterChainMatch = FilterChainMatch.newBuilder()
+        .setSourceType(FilterChainMatch.ConnectionSourceType.ANY)
+        .build();
+    FilterChain filterChain = FilterChain.newBuilder()
+        .setName("filter-chain-0")
+        .setFilterChainMatch(filterChainMatch)
+        .addFilters(filter)
+        .build();
+    return Listener.newBuilder()
+        .setName(name)
+        .setTrafficDirection(TrafficDirection.INBOUND)
+        .addFilterChains(filterChain)
+        .build();
+  }
+
+  private static RouteConfiguration rds(String authority) {
+    VirtualHost virtualHost = VirtualHost.newBuilder()
+        .addDomains(authority)
+        .addRoutes(
+            Route.newBuilder()
+                .setMatch(
+                    RouteMatch.newBuilder().setPrefix("/").build())
+                .setRoute(
+                    RouteAction.newBuilder().setCluster(clusterName).build()).build()).build();
+    return RouteConfiguration.newBuilder().setName(rdsName).addVirtualHosts(virtualHost).build();
+  }
+
+  private static Cluster cds() {
+    return Cluster.newBuilder()
+        .setName(clusterName)
+        .setType(Cluster.DiscoveryType.EDS)
+        .setEdsClusterConfig(
+            Cluster.EdsClusterConfig.newBuilder()
+                .setServiceName(edsName)
+                .setEdsConfig(
+                    ConfigSource.newBuilder()
+                        .setAds(AggregatedConfigSource.newBuilder().build())
+                        .build())
+                .build())
+        .setLbPolicy(Cluster.LbPolicy.ROUND_ROBIN)
+        .build();
+  }
+
+  private static ClusterLoadAssignment eds(int port) {
+    Address address = Address.newBuilder()
+        .setSocketAddress(
+            SocketAddress.newBuilder().setAddress("0.0.0.0").setPortValue(port).build()).build();
```

Comment:
I'm surprised this works. I'd expect the IP address 127.0.0.1 or ::1 here.

Reply:
oh no I somehow remembered it wrong that the loopback address 0.0.0.0 and 127.0.0.1 was equivalent.....
However, it does work! I double checked the log, we got the golden respond and the right eds...
it also works with `127.0.0.1` or `::1` but not something like `0.0.0.1` or `1.2.3.4`...
```
 io.grpc.netty.shaded.io.netty.channel.AbstractChannel$AnnotatedNoRouteToHostException: No route to host: /0.0.0.1:59960
```

Reply:
`0.0.0.0` in Unix APIs is generally a "wildcard IP." It is used commonly on server-side to listen on all IPs for the host. But apparently it can also be used as a destination address synonymous with localhost: https://superuser.com/a/1307057. But apparently [it can still break sometimes](https://stackoverflow.com/q/54636989), for unknown reasons?

In any case, seems like it'd be best to just use `InetAddress.getLoopbackAddress()` or `server.getListenSockets().get(0)`. That's a bit annoying if you need to form a "host:port" string, but it seems we don't need to form such a string.